### PR TITLE
Load chain config from file

### PR DESCRIPTION
**Pull request by shayzluf**
Feature

**What does this PR do? Why is it needed?**
load chain config from yaml file
**Which issues(s) does this PR fix?**

Part of #5693

### CODE REVIEW

File: shared/cmd/flags.go [hunk 1]
```diff
@@ -171,7 +171,12 @@ var (
 	}
 	// ConfigFileFlag specifies the filepath to load flag values.
 	ConfigFileFlag = &cli.StringFlag{
-		Name:  "config-file",
+		Name:  "flag-config-file",
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply shared/cmd/flags.go:174]: was wandering about it. thanks

File: shared/cmd/flags.go [hunk 2]
```diff
@@ -174,4 +174,9 @@ var (
 		Name:  "config-file",
 		Usage: "The filepath to a yaml file with flag values",
 	}
+	// ChainConfigFileFlag specifies the filepath to load flag values.
+	ChainConfigFileFlag = &cli.StringFlag{
+		Name:  "chain-config-file",
+		Usage: "The filepath to a yaml file with chain config values",
```
Comment (farazdagi) [inline shared/cmd/flags.go:180]:
```suggestion
		Usage: "The path to a YAML file with chain config values",
```

File: beacon-chain/main.go [hunk 3]
```diff
@@ -141,7 +146,18 @@ func main() {
...
-
+		if ctx.IsSet(cmd.ChainConfigFileFlag.Name) {
+			chainConfigFileName := ctx.String(cmd.ChainConfigFileFlag.Name)
+			yamlFile, err := ioutil.ReadFile(chainConfigFileName)
+			if err != nil {
+				log.WithError(err).Error("Failed to read chain config file.")
+			}
+			conf := params.BeaconConfig()
+			if err := yaml.Unmarshal(yamlFile, conf); err != nil {
+				log.WithError(err).Error("Failed to parse chain config yaml file.")
```
Comment (farazdagi) [inline beacon-chain/main.go:157]:
```suggestion
				log.WithError(err).Error("Failed to parse chain config YAML file.")
```
YAML is acronym, so should be in all caps.

File: beacon-chain/main.go [hunk 4]
```diff
@@ -3,10 +3,13 @@ package main
 
 import (
 	"fmt"
+	"io/ioutil"
 	"os"
 	"runtime"
 	runtimeDebug "runtime/debug"
 
+	"gopkg.in/yaml.v2"
+
```
Comment (farazdagi) [inline beacon-chain/main.go:12]: consider moving this import down (it stands in its own group), and go fmt.

File: shared/params/config.go [hunk 5]
```diff
@@ -16,9 +16,9 @@ type BeaconChainConfig struct {
...
-	MaxCommitteesPerSlot           uint64 // MaxCommitteesPerSlot defines the max amount of committee in a single slot.
+	TargetCommitteeSize            uint64 `yaml:"TARGET_COMMITTEE_SIZE"`              // TargetCommitteeSize is the number of validators in a committee when the chain is healthy.
+	MaxValidatorsPerCommittee      uint64 `yaml:"MAX_VALIDATORS_PER_COMMITTEE"`       // MaxValidatorsPerCommittee defines the upper bound of the size of a committee.
+	MaxCommitteesPerSlot           uint64 `yaml:"MAX_COMMITTEES_PER_SLOT"`            // MaxCommitteesPerSlot defines the max amount of committee in a single slot.
 	MinPerEpochChurnLimit          uint64 `yaml:"MIN_PER_EPOCH_CHURN_LIMIT"`          // MinPerEpochChurnLimit is the minimum amount of churn allotted for validator rotations.
 	ChurnLimitQuotient             uint64 `yaml:"CHURN_LIMIT_QUOTIENT"`               // ChurnLimitQuotient is used to determine the limit of how many validators can rotate per epoch.
 	ShuffleRoundCount              uint64 `yaml:"SHUFFLE_ROUND_COUNT"`                // ShuffleRoundCount is used for retrieving the permuted index.
```
Comment (terencechain) [inline shared/params/config.go:24]: Have you have a chance to check all the yaml tags are correct based on the latest spec?

I'm happy to check as well, just wanted to know if it was done already : )

Comment (shayzluf) [reply shared/params/config.go:24]: planned to do it in another PR part of #5693

File: beacon-chain/main.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/main.go]: thanks, fixed

**Comment by shayzluf** [comment]
@nisdas will handle both issues. thanks

**Comment by shayzluf** [comment]
@nisdas handled both issues:
- convert 0x to fixed byte arrays
- load default values before un marshal

File: beacon-chain/main.go [hunk 6]
```diff
@@ -163,6 +173,82 @@ func main() {
 	}
 }
 
+func loadChainConfigFile(chainConfigFileName string) {
```
Comment (nisdas) [inline beacon-chain/main.go:176]: why is this in `main.go` instead of the `node` package ?

Comment (shayzluf) [reply beacon-chain/main.go]: moved

File: beacon-chain/node/node.go [hunk 7]
```diff
@@ -662,3 +669,79 @@ func (b *BeaconNode) registerArchiverService() error {
...
+			case l > 32 && l <= 48:
+				var arr [48]byte
+				copy(arr[:], b)
+				fixedByte, err := yaml.Marshal(arr)
+				if err != nil {
+					log.WithError(err).Error("Failed to marshal config file.")
+				}
+				parts[1] = string(fixedByte)
+			case l > 48 && l <= 64:
+				var arr [48]byte
```
Comment (nisdas) [inline beacon-chain/node/node.go:721]: change to 64

File: beacon-chain/node/node.go [hunk 8]
```diff
@@ -662,3 +669,79 @@ func (b *BeaconNode) registerArchiverService() error {
...
+				copy(arr[:], b)
+				fixedByte, err := yaml.Marshal(arr)
+				if err != nil {
+					log.WithError(err).Error("Failed to marshal config file.")
+				}
+				parts[1] = string(fixedByte)
+			case l > 64 && l <= 96:
+				var arr [48]byte
```
Comment (nisdas) [inline beacon-chain/node/node.go:729]: change to 96

File: beacon-chain/node/node.go [hunk 9]
```diff
@@ -662,3 +669,79 @@ func (b *BeaconNode) registerArchiverService() error {
...
 	return b.services.RegisterService(svc)
 }
+
+func loadChainConfigFile(chainConfigFileName string) {
+	yamlFile, err := ioutil.ReadFile(chainConfigFileName)
+	if err != nil {
+		log.WithError(err).Error("Failed to read chain config file.")
+	}
+	// convert 0x hex inputs to fixed bytes arrays
+	lines := strings.Split(string(yamlFile), "\n")
```
Comment (nisdas) [inline beacon-chain/node/node.go:679]: can you refactor this whole chunk into a separate method and unit test it

File: beacon-chain/node/node.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/node/node.go]: fixed, thanks

File: beacon-chain/node/node.go (hunk 8 above)
Comment (shayzluf) [reply beacon-chain/node/node.go]: fixed

File: beacon-chain/node/node.go (hunk 9 above)
Comment (shayzluf) [reply beacon-chain/node/node.go:679]: done, thanks

File: beacon-chain/main.go [hunk 10]
```diff
@@ -25,6 +25,7 @@ import (
 	"gopkg.in/urfave/cli.v2/altsrc"
 )
 
+var log *logrus.Entry
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply beacon-chain/main.go]: sure reverted

File: beacon-chain/node/node.go [hunk 11]
```diff
@@ -684,6 +684,8 @@ func loadChainConfigFile(chainConfigFileName string) {
 		}
 	}
 	yamlFile = []byte(strings.Join(lines, "\n"))
+	lines = strings.Split(string(yamlFile), "\n")
+	log.Info(lines)
```
Comment (nisdas) [inline beacon-chain/node/node.go:688]: why are you logging this ?

Comment (shayzluf) [reply beacon-chain/node/node.go:688]: sorry , for debugging

Comment (rauljordan) [reply beacon-chain/node/node.go:688]: Should be good to remove, likely was for debugging

File: beacon-chain/node/node.go [hunk 12]
```diff
@@ -662,3 +669,111 @@ func (b *BeaconNode) registerArchiverService() error {
...
+	// convert 0x hex inputs to fixed bytes arrays
+	lines := strings.Split(string(yamlFile), "\n")
+	for i, line := range lines {
+		if !strings.HasPrefix(line, "#") && strings.Contains(line, "0x") {
+			parts := replaceHexStringWithYAMLFormat(line)
+			lines[i] = strings.Join(parts, "\n")
+		}
+	}
+	yamlFile = []byte(strings.Join(lines, "\n"))
+	lines = strings.Split(string(yamlFile), "\n")
```
Comment (rauljordan) [inline beacon-chain/node/node.go:687]:
```suggestion
```

File: beacon-chain/node/node.go [hunk 13]
```diff
@@ -662,3 +669,111 @@ func (b *BeaconNode) registerArchiverService() error {
...
+	log.Info(lines)
```
Comment (rauljordan) [inline beacon-chain/node/node.go:688]:
```suggestion
```

File: beacon-chain/node/node.go [hunk 14]
```diff
@@ -662,3 +669,109 @@ func (b *BeaconNode) registerArchiverService() error {
...
+			parts := replaceHexStringWithYAMLFormat(line)
+			lines[i] = strings.Join(parts, "\n")
+		}
+	}
+	yamlFile = []byte(strings.Join(lines, "\n"))
+	conf := params.BeaconConfig()
+	if err := yaml.Unmarshal(yamlFile, conf); err != nil {
+		log.WithError(err).Fatal("Failed to parse chain config yaml file.")
+	}
+	log.Infof("Config file values: %+v", conf)
```
Comment (nisdas) [inline beacon-chain/node/node.go:691]: I dont think we should be logging this

Comment (shayzluf) [reply beacon-chain/node/node.go:691]: should i remove it?

Comment (nisdas) [reply beacon-chain/node/node.go:691]: yes, its going to be noisy and confusing for the average user.

Comment (shayzluf) [reply beacon-chain/node/node.go]: changed to debug

File: beacon-chain/node/node.go [hunk 15]
```diff
@@ -662,3 +669,109 @@ func (b *BeaconNode) registerArchiverService() error {
 	})
 	return b.services.RegisterService(svc)
 }
+
+func loadChainConfigFile(chainConfigFileName string) {
+	yamlFile, err := ioutil.ReadFile(chainConfigFileName)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to read chain config file.")
+	}
+	// convert 0x hex inputs to fixed bytes arrays
```
Comment (terencechain) [inline beacon-chain/node/node.go:678]:
```suggestion
	// Convert 0x hex inputs to fixed bytes arrays.
```

File: beacon-chain/node/node.go [hunk 16]
```diff
@@ -662,3 +669,109 @@ func (b *BeaconNode) registerArchiverService() error {
 	})
 	return b.services.RegisterService(svc)
 }
+
+func loadChainConfigFile(chainConfigFileName string) {
```
Comment (terencechain) [inline beacon-chain/node/node.go:673]: Can `loadChainConfigFile` and `replaceHexStringWithYAMLFormat` not inside `node.go`?

These are only helper methods just for loading custom chain configs.

Comment (shayzluf) [reply beacon-chain/node/node.go]: moved

**Pull request merged**